### ISSUE TITLE
Resource linking fails in projects with multiple modules. 

### ISSUE THREAD

**Issue by xoif** [AllBus/gradle-android-scala-plugin#5]
In projects with multiple modules resource linking is failing for modules written in scala. Although the project will build without any errors this will lead to crashes at runtime. 
Our project combines a module completely written in scala with several other modules written in java and/or kotlin.
As another user just recently reported a similar [problem](https://github.com/AllBus/gradle-android-scala-plugin/issues/2#issuecomment-553847692), I decided to open a new issue as I'm not able to re-open the [old one](https://github.com/AllBus/gradle-android-scala-plugin/issues/2#issue-434185862). 

The crash happens as soon as the first resource (normally the content view) is referenced. 
`11-18 10:09:05.804 7917-7917/<package name> E/class: uncaught exception: java.lang.RuntimeException: Unable to start activity ComponentInfo{<package name>.ExampleActivity}: android.content.res.Resources$NotFoundException: Resource ID #0x7f0e0038
11-18 10:09:06.279 7917-7917/<package name> E/AndroidRuntime: FATAL EXCEPTION: main
    Process: <package name>, PID: 7917
    java.lang.RuntimeException: Unable to start activity ComponentInfo{<package name>.ExampleActivity}: android.content.res.Resources$NotFoundException: Resource ID #0x7f0e0038`

linked Resource ID in R.txt of scala module: 
`int layout activity_example_main 0x7f0e0038`

same resource in _application module_/build/intermediates/symbols/debug/R.txt:
`int layout activity_example_main 0x7f0c0048`

respective generated R.java file of _application module_/build/generated/not_namespaced_r_class_sources/debug/r/_scala submodule_/R.java:
`public static final int activity_example_main = 0x7f0c0048;`

**Comment by mx-tc** [comment]
By lots of trial and error I came to the following solution to avoid letting Scala-code directly use any of the integers behind those R.java constants. We created a java-proxy-class that provides access to the resource-ids through java-reflection:

```
import androidx.annotation.NonNull;
import java.lang.reflect.Field;

public class ScalaCompatR {
    private static <T> int get(@NonNull Class<T> clazz, @NonNull String resourceName) {
        try {
            Field f = clazz.getDeclaredField(resourceName);
            return f.getInt(f);
        } catch (NoSuchFieldException | IllegalAccessException e) {
            e.printStackTrace();
            return 0;
        }
    }

    public static int getLayout(@NonNull String resourceName) {
        return get(R.layout.class, resourceName);
    }

    public static int getId(@NonNull String resourceName) {
        return get(R.id.class, resourceName);
    }

    public static int getDrawable(@NonNull String resourceName) {
        return get(R.drawable.class, resourceName);
    }

    public static int getString(@NonNull String resourceName) {
        return get(R.string.class, resourceName);
    }

    public static int getColor(@NonNull String resourceName) {
        return get(R.color.class, resourceName);
    }

    public static int getAnim(@NonNull String resourceName) {
        return get(R.anim.class, resourceName);
    }

    public static int getAnimator(@NonNull String resourceName) {
        return get(R.animator.class, resourceName);
    }

    public static int getPlurals(@NonNull String resourceName) {
        return get(R.plurals.class, resourceName);
    }

    public static int getMenu(@NonNull String resourceName) {
        return get(R.menu.class, resourceName);
    }

    public static int getDimen(@NonNull String resourceName) {
        return get(R.dimen.class, resourceName);
    }
}
```

We than replaced access to resource-ids in Scala-code by something like

```
  override def onCreate(bundle:Bundle) = {
    super.onCreate(bundle)
    setContentView(ScalaCompatR.getLayout("activity_example"))
  }
```

This way the runtime-exception seems to be fixed as the proper resource-ids get used to look up the resources within the APK.
Maybe this helps in finding another more clean solution of this problem.

**Comment by AllBus** [comment]
In my case for access to R resources without reflection:
Create java proxy class in module (in the same scala module or in another module)

```
public class ProxyScalaR {
    public static final class string {
        public static final int bravo = R.string.bravo;
    }
}
```

And use this in scala classes as

```
 def getText(context: Context): String ={
    context.getString(ProxyScalaR.string.bravo)
 }
```

**Comment by AllBus** [comment]
I’m exploring how to solve this problem at the plugin level.

**Comment by AllBus** [comment]
Example project https://github.com/AllBus/ScalaMultiModuleAndroidPluginExample

**Comment by xoif** [comment]
As both suggestions come with drawbacks, none of them are really suitable for production builds. @AllBus is there already any progress on your attempt to fix this bug at the plugin level?

**Comment by AllBus** [comment]
The problem is that the scala compiler inlined R constants to class-files.
It is necessary to find a way in which in Scala classes compiler will save the full paths to R values in Scala classes values like for Java classes.

**Comment by AllBus** [comment]
As an experiment, you can add a dependency to scala module on the app module R jar

for example:

   ` implementation fileTree("../app/build/intermediates/compile_and_runtime_not_namespaced_r_class_jar/debug/R.jar")`

**Comment by mx-tc** [comment]
I doubt this is caused due to "illegal" inlining the R constant-values by the scala-compiler (although there seems to exist such a mechanism, which can get observed here, see below)

### First some background-information I found after some investigation:

1. Since [Android Gradle Plugin 3.3.0](https://developer.android.com/studio/releases/gradle-plugin#3-3-0) the processing of resources directly lead to a file below `build/intermediates/compile_only_not_namespaced_r_class_jar/debug/R.jar`.
> Previously, the Android Gradle plugin would generate an R.java file for each of your project's dependencies and then compile those R classes alongside your app's other classes. The plugin now generates a JAR containing your app's compiled R class directly, without first building intermediate R.java classes.

This JAR in turn doesn't have any R-constant-*values* - just the field declarations (as can get evaluated by extracting the JAR and disassembling the contained class-files using e.g. `java -p -c R$string.class`) 
So if during the scala-compile-run only this JAR-file would be added to the classpath, it should be impossible to inline any constant-values in any way...

2. But somehow also a file `build/generated/not_namespaced_r_class_sources/debugAndroidTest/r/<package-path>/R.java` gets generated during build, which in turn has any of the R-constants defined using `static final` modifiers! As can get read in many internet-threads those fields are valid for getting inlined even by the java-compiler (some nice reading about this can get found [here](https://www.javaworld.com/article/2072985/inconstant-constants-in-java.html)). 
So to me it seems as if during the scala-compile-run this file `R.java` is "seen" by the scala-compiler, which then is allowed to get inlined its values like the java-compiler does as well.

### Experiments about Scala inlining of Java-fields
To be sure about inlining of Java-fields in Scala I used a very basic Android-project (similar to the one linked here before).
I then added two classes to the scala-submodule:

- Java-file `TestR.java`, which mimics the normal R-file but *without* having its fields declared as `final`
```
package com.example.mylibrary;

public final class TestR {
    private TestR() { }
    public static final class string {
        private string() { }
        public static int app_name = 0x7f0b001b;
    }
}
```

- Scala-file `ScalaDummy.scala`, accessing one regular R-constant as well as a field of `TestR`
```
package com.example.mylibrary

import android.util.Log

class ScalaDummy {
  def testString(): Unit = {
    Log.d("ScalaTest", "R.string.app_name-id: " + R.string.app_name)
    Log.d("ScalaTest", "TestR.string.app_name-id: " + TestR.string.app_name)
  }
}
```

After having built the project I then decompiled the according class-file from `build/intermediates/javac/debug/classes/com/example/mylibrary/ScalaDummy.class` (again using `javap -p -c ScalaDummy.class`)
```
Compiled from "ScalaDummy.scala"
public class com.example.mylibrary.ScalaDummy {
  public void testString();
    Code:
       0: ldc           #12                 // String ScalaTest
       2: new           #14                 // class scala/collection/mutable/StringBuilder
       5: dup
       6: invokespecial #17                 // Method scala/collection/mutable/StringBuilder."<init>":()V
       9: ldc           #19                 // String R.string.app_name-id:
      11: invokevirtual #23                 // Method scala/collection/mutable/StringBuilder.append:(Ljava/lang/Object;)Lscala/collection/mutable/StringBuilder;
      14: ldc           #24                 // int 2132017180
      16: invokestatic  #30                 // Method scala/runtime/BoxesRunTime.boxToInteger:(I)Ljava/lang/Integer;
      19: invokevirtual #23                 // Method scala/collection/mutable/StringBuilder.append:(Ljava/lang/Object;)Lscala/collection/mutable/StringBuilder;
      22: invokevirtual #34                 // Method scala/collection/mutable/StringBuilder.toString:()Ljava/lang/String;
      25: invokestatic  #40                 // Method android/util/Log.d:(Ljava/lang/String;Ljava/lang/String;)I
      28: pop
      29: ldc           #12                 // String ScalaTest
      31: new           #14                 // class scala/collection/mutable/StringBuilder
      34: dup
      35: invokespecial #17                 // Method scala/collection/mutable/StringBuilder."<init>":()V
      38: ldc           #42                 // String TestR.string.app_name-id:
      40: invokevirtual #23                 // Method scala/collection/mutable/StringBuilder.append:(Ljava/lang/Object;)Lscala/collection/mutable/StringBuilder;
      43: getstatic     #48                 // Field com/example/mylibrary/TestR$string.app_name:I
      46: invokestatic  #30                 // Method scala/runtime/BoxesRunTime.boxToInteger:(I)Ljava/lang/Integer;
      49: invokevirtual #23                 // Method scala/collection/mutable/StringBuilder.append:(Ljava/lang/Object;)Lscala/collection/mutable/StringBuilder;
      52: invokevirtual #34                 // Method scala/collection/mutable/StringBuilder.toString:()Ljava/lang/String;
      55: invokestatic  #40                 // Method android/util/Log.d:(Ljava/lang/String;Ljava/lang/String;)I
      58: pop
      59: return

  public com.example.mylibrary.ScalaDummy();
    Code:
       0: aload_0
       1: invokespecial #51                 // Method java/lang/Object."<init>":()V
       4: return
}
```

So the first call of `Log.d()` accessing `R.string.app_name` leads to the inlined integer-value (see decompiled code-line 14, opcode `ldc`), but the other call accessing `TestR.string.app_name` shows linking to the static field, as it should be (code-line 43, opcode `getstatic`).

### Long story short: 
As already mentioned above it seems as if accidentally the file `R.java` gets compiled during building the scala-submodule instead of just compiling the scala-files and linking in the `R.jar`.

I hope my very lengthy descriptions of my investigations can help identify the real cause of the problem and enables you to fix the issue. Maybe the solution is to somehow remove the generated `R.java` from the source-sets.

**Comment by AllBus** [comment]
The question arises: why, in the case of the R class, the value is inlined, while for TestR it refers to the static field?

**Comment by AllBus** [comment]
you can see source set via gradle task

```
gradle.taskGraph.whenReady {taskGraph ->

    println "Found task graph: " + taskGraph

    taskGraph.allTasks.forEach { task ->
        println()
        println("----- " + task + " -----")

        println("inputs: ")
        task.inputs.getFiles().getFiles().collect { f -> println(" - " + f)}
        println("outputs: ")
        task.outputs.getFiles().getFiles().collect { f -> println(" + " + f)}
    }
}

```

You can also make sure that `R.java` not exists in generated code on lastversion build tools

**Comment by AllBus** [comment]
It two jars in source set which contains R.class

```
module\build\intermediates\compile_only_not_namespaced_r_class_jar\debug\R.jar
~\.gradle\caches\transforms-2\files-2.1\d4ddef2c22294d5122f8ec03f8a6621f\jetified-R.jar
```

First without defined values, second contains resources values

**Comment by AllBus** [comment]
I create tiny scala project (not android) contain one Object and two dependencies: `R.jar` from android module and `RM.jar` which contains

```
package com.kos.sacamodule;

public final class R {
    private R() {
    }

    public static final class anim {
        public static int abc_fade_in;
        public static int abc_fade_out;
        public static int abc_grow_fade_in_from_bottom;
    }
}


``` 

The project object:


```
object Hello extends App {
  println("My compile "+com.kos.sacamodule.R.anim.abc_grow_fade_in_from_bottom+" ok.")
  println("From android "+com.kos.javamodule.R.anim.abc_grow_fade_in_from_bottom+" hm.")
}
```


It compiled to 

 ```
 0: getstatic     #61                 // Field scala/Predef$.MODULE$:Lscala/Predef$;
   3: new           #63                 // class scala/collection/mutable/StringBuilder
   6: dup
   7: invokespecial #64                 // Method scala/collection/mutable/StringBuilder."<init>":()V
  10: ldc           #66                 // String My compile
  12: invokevirtual #70                 // Method scala/collection/mutable/StringBuilder.append:(Ljava/lang/Object;)Lscala/collection/mutable/StringBuilder;
  15: getstatic     #76                 // Field com/kos/sacamodule/R$anim.abc_grow_fade_in_from_bottom:I
  18: invokestatic  #82                 // Method scala/runtime/BoxesRunTime.boxToInteger:(I)Ljava/lang/Integer;
  21: invokevirtual #70                 // Method scala/collection/mutable/StringBuilder.append:(Ljava/lang/Object;)Lscala/collection/mutable/StringBuilder;
  24: ldc           #84                 // String  ok.
  26: invokevirtual #70                 // Method scala/collection/mutable/StringBuilder.append:(Ljava/lang/Object;)Lscala/collection/mutable/StringBuilder;
  29: invokevirtual #88                 // Method scala/collection/mutable/StringBuilder.toString:()Ljava/lang/String;
  32: invokevirtual #92                 // Method scala/Predef$.println:(Ljava/lang/Object;)V
  35: getstatic     #61                 // Field scala/Predef$.MODULE$:Lscala/Predef$;
  38: new           #63                 // class scala/collection/mutable/StringBuilder
  41: dup
  42: invokespecial #64                 // Method scala/collection/mutable/StringBuilder."<init>":()V
  45: ldc           #94                 // String From android
  47: invokevirtual #70                 // Method scala/collection/mutable/StringBuilder.append:(Ljava/lang/Object;)Lscala/collection/mutable/StringBuilder;
  50: iconst_0
  51: invokestatic  #82                 // Method scala/runtime/BoxesRunTime.boxToInteger:(I)Ljava/lang/Integer;
  54: invokevirtual #70                 // Method scala/collection/mutable/StringBuilder.append:(Ljava/lang/Object;)Lscala/collection/mutable/StringBuilder;
  57: ldc           #96                 // String  hm.
  59: invokevirtual #70                 // Method scala/collection/mutable/StringBuilder.append:(Ljava/lang/Object;)Lscala/collection/mutable/StringBuilder;
  62: invokevirtual #88                 // Method scala/collection/mutable/StringBuilder.toString:()Ljava/lang/String;
  65: invokevirtual #92                 // Method scala/Predef$.println:(Ljava/lang/Object;)V
  68: return
```

line `15: getstatic` - it from my compiled RM.jar
line `50: iconst_0` - if from android R.jar


Thus, inlining is independent of android
It seems worth trying to create own `R.java`
